include/execution.hpp

```diff
@@ -2852,7 +2852,217 @@ namespace _P2300::execution {
   /////////////////////////////////////////////////////////////////////////////
   // [execution.senders.adaptors.ensure_started]
   namespace __ensure_started {
+
+    template <class _SharedState>
+      class __receiver {
+        _SharedState& __shared_state_;
+
+      public:
+        explicit __receiver(_SharedState &__shared_state) noexcept
+          : __shared_state_(__shared_state) {
+        }
+
+        template <__one_of<set_value_t, set_error_t, set_stopped_t> _Tag, class... _As>
+        friend void tag_invoke(_Tag __tag, __receiver&& __self, _As&&... __as) noexcept {
+          _SharedState& __state = __self.__shared_state_;
+
+          try {
+            using __tuple_t = __decayed_tuple<_Tag, _As...>;
+            __state.__data_.template emplace<__tuple_t>(__tag, (_As &&) __as...);
+          } catch (...) {
+            using __tuple_t = __decayed_tuple<set_error_t, std::exception_ptr>;
+            __state.__data_.template emplace<__tuple_t>(set_error, std::current_exception());
+          }
+
+          __state.__notify();
+        }
+
+        friend auto tag_invoke(get_env_t, const __receiver& __self)
+          -> make_env_t<with_t<get_stop_token_t, in_place_stop_token>> {
+          return make_env(with(get_stop_token, __self.__shared_state_.__stop_source_.get_token()));
+        }
+      };
+
+    struct __operation_base {
+      using __notify_fn = void(__operation_base*) noexcept;
+      __notify_fn* __notify_{};
+    };
+
+    template <class _SenderId>
+      struct __sh_state {
+        using _Sender = __t<_SenderId>;
+
+        template <class... _Ts>
+          using __bind_tuples =
+            __mbind_front_q<
+              __variant,
+              std::tuple<set_stopped_t>, // Initial state of the variant is set_stopped
+              std::tuple<set_error_t, std::exception_ptr>,
+              _Ts...>;
+
+        using __bound_values_t =
+          __value_types_of_t<
+            _Sender,
+            make_env_t<with_t<get_stop_token_t, in_place_stop_token>>,
+            __mbind_front_q<__decayed_tuple, set_value_t>,
+            __q<__bind_tuples>>;
+
+        using __variant_t =
+          __error_types_of_t<
+            _Sender,
+            make_env_t<with_t<get_stop_token_t, in_place_stop_token>>,
+            __transform<
+              __mbind_front_q<__decayed_tuple, set_error_t>,
+              __bound_values_t>>;
+
+        using __receiver = __receiver<__sh_state>;
+
+        __variant_t __data_;
+        in_place_stop_source __stop_source_{};
+
+        std::atomic<void*> __op_state1_;
+        connect_result_t<_Sender, __receiver> __op_state2_;
+
+        explicit __sh_state(_Sender& __sndr)
+          : __op_state1_{nullptr}
+          , __op_state2_(connect((_Sender&&) __sndr, __receiver{*this}))
```

Comment:
Ignoring the question of whether `ensure_started` should exist in the first place, `__receiver` also needs to increment the reference count of the shared state to be able to support the case where the `ensure_started` sender is destroyed without ever connecting it. Otherwise the sender holds the only (reference counted) reference to the shared state, and it'll likely be destroyed before the receiver is signalled. This requirement definitely convolutes the implementation and I'm not quite sure how to do it with a `shared_ptr`, but in HPX and pika we've done it relatively cleanly with an intrusive pointer.